awesomerc.lua

```diff
@@ -372,6 +372,33 @@ awful.keyboard.append_global_keybindings({
                 end
             end
         end,
+    },
+    awful.key {
```

Comment:
The concept itself is fine, I did that for many years. They way this is implemented isn't really gonna work in practice. I mean, technically "works" for the default scenario, but many people don't use layouts that way. Also, when the new layout engine lands, without further abstraction layer(s), this will blow up.

### What are my concerns with this

* If the user doesn't have a numpad, this doesn't work. Most keyboard these days don't.
* It adds a lot of imperative code to `rc.lua` while we try very hard to remove all imperative code.
* Some international keyboard layouts have different key name.
* Numlock will mess this up.
* Individual tags can have their own layout list, this will conflict.
* The `awful.widget.layoutlist` is based either on the global set of layouts or the individual tag ones. If the user use that widget (for an alt-tab style popup or in the wibar), using this mechanism will cause a lot of fallback code to kick in. This fallback code has unwanted side effects, it should not be possible to enter these code paths using the default config.
* This bypasses the `append_default_layout`/`remove_default_layout`. The reason those exist is because they allow self-contained modules to add their own layouts. In the long term, I am removing all "write once" APIs in favor of read/write/append/clear ones. The reason for this is modularity.

Now, the better news: I agree we should add this feature, it is useful. However, if you want this merged, you will have to take the long way to get there.

### Expand the C code API to convert the keysym and keynames correctly in a way that universially works

We have to maintain this code going forward. History has proven time and again that if such hardcoding make its way into actual business logic (ie, not the hotkey popup, which is decoupled for the core), people will report bugs. This is why I added the C code for the modifiers, I did this hack and it blew up in my face.

This C code will need a minimal test suite. Don't worry too much about that, integration tests for such thing are really trivial to add (I don't mind doing it if you have no time).

Please do so in its own pull request, that's would be a rather nice and uncontroversial addition and pave the way for "what you really want".

### Use the current tag `layout` property instead of the array

As mentioned above, there is very valid reasons to not hardcode this in `rc.lua`.

* It is not compatible with self-contained modules
* It has side effects when the `layoutlist` widget is used
* It triggers some fallback code in `lib/awful/tag.lua`
* It disregard user defined tag list. Those list will soon be easier to setup with I merge `ruled.tag`.
* When the new layout module lands, people will start to use it and it will conflict with this hardcoded logic.

By using `awful.screen.focused().selected_tag.layouts[index]`, you could mitigate most of those concerns. It would also keep the size overhead to a minimum.

Reply:
Attempting to expand the API at #3011.